src/java.base/share/native/libjli/java.c

```diff
@@ -1372,6 +1372,8 @@ ParseArguments(int *pargc, char ***pargv,
     }
 
     if (mode == LM_SOURCE) {
+        // signal module bootstrap to defer warnings about unknown modules
+        AddOption("-Djdk.internal.java.launchmode=source", NULL);
```

Comment:
This is the launcher code so better if the comment said could say that it communicates the launcher mode to runtime (confusing to say "signal module bootstrap" or anything about warnings here).

I note that the existing property is "sun.java.launcher". If you want then you could continue this contention and use "sun.java.launcher.mode".  It'a all JDK-internal and the system property is removed during startup so the naming doesn't really matter of course, I'd like prefer it end in ".mode" rather than "launchmode".